tools/oeedger8r/Emitter.ml

```diff
@@ -325,153 +380,139 @@ let oe_gen_args_header (ec : enclave_content) (dir : string) =
+  ; "memcpy(_pargs_in, &_args, sizeof(*_pargs_in));" ]
 
-let oe_process_output_buffer (os : out_channel) (fd : func_decl) =
-  (* Verify that the ecall succeeded *)
-  fprintf os "    /* Set up output arg struct pointer */\n" ;
-  fprintf os "    _pargs_out = (%s_args_t*)_output_buffer;\n" fd.fname ;
-  fprintf os "    OE_ADD_SIZE(_output_buffer_offset, sizeof(*_pargs_out));\n\n" ;
-  fprintf os "    /* Check if the call succeeded */\n" ;
-  fprintf os "    if ((_result=_pargs_out->_result) != OE_OK)\n" ;
-  fprintf os "        goto done;\n\n" ;
-  fprintf os
-    "    /* Currently exactly _output_buffer_size bytes must be written */\n" ;
-  fprintf os "    if (_output_bytes_written != _output_buffer_size) {\n" ;
-  fprintf os "        _result = OE_FAILURE;\n" ;
-  fprintf os "        goto done;\n" ;
-  fprintf os "    }\n\n" ;
-  (* Unmarshal return value and ouput buffers *)
-  fprintf os "    /* Unmarshal return value and out, in-out parameters */\n" ;
-  if fd.rtype <> Void then fprintf os "    *_retval = _pargs_out->_retval;\n" ;
-  List.iter
-    (fun (ptype, decl) ->
-      match ptype with
-      | PTPtr (atype, ptr_attr) ->
-          if ptr_attr.pa_chkptr then
-            let size = oe_get_param_size (ptype, decl, "_args.") in
-            match ptr_attr.pa_direction with
-            | PtrOut ->
-                (* strings cannot be out parameters *)
-                fprintf os "    OE_READ_OUT_PARAM(%s, (size_t)(%s));\n"
-                  decl.identifier size
-            | PtrInOut ->
-                (* Check that strings are null terminated.
-                 Note output buffer has already been copied into the enclave.*)
-                if ptr_attr.pa_isstr then
-                  fprintf os
-                    "    OE_CHECK_NULL_TERMINATOR(_output_buffer + \
-                     _output_buffer_offset, _args.%s_len);\n"
-                    decl.identifier
-                else if ptr_attr.pa_iswstr then
-                  fprintf os
-                    "    OE_CHECK_NULL_TERMINATOR_WIDE(_output_buffer + \
-                     _output_buffer_offset, _args.%s_len);\n"
-                    decl.identifier
-                else () ;
-                fprintf os "    OE_READ_IN_OUT_PARAM(%s, (size_t)(%s));\n"
-                  decl.identifier size
-            | _ -> ()
-          else ()
-      | _ -> () )
-    fd.plist ;
-  fprintf os "\n"
+let oe_process_output_buffer (fd : func_decl) =
+  List.flatten
+    [ [ (* Verify that the ecall succeeded *)
+        ""
+      ; "/* Setup output arg struct pointer. */"
+      ; sprintf "_pargs_out = (%s_args_t*)_output_buffer;" fd.fname
+      ; "OE_ADD_SIZE(_output_buffer_offset, sizeof(*_pargs_out));"
+      ; ""
+      ; "/* Check if the call succeeded. */"
+      ; "if ((_result = _pargs_out->_result) != OE_OK)"
+      ; "    goto done;"
+      ; ""
+      ; "/* Currently exactly _output_buffer_size bytes must be written. */"
```

Comment:
In the future we can tweaks this to support variable length outputs.

Reply:
That'd be cool.